# 第2章 章末演習問題 (全61問)

教科書「パターン認識と機械学習 (PRML)」第2章「確率分布」のすべての Exercises (2.1 〜 2.61) を網羅しています。
証明・論述問題については導出のステップと論理展開を提示した上で重要な数式・語句を穴埋め（`___`）形式としています。
計算・シミュレーション問題については Python コードセル（`None` や `# YOUR CODE HERE`）を配置しています。

## Exercise 2.1 (ベルヌーイ分布の規格化、モーメント、エントロピー)
ベルヌーイ分布 $p(x|\mu) = \mu^x (1 - \mu)^{1 - x}$ ($x \in \{0, 1\}$) について、以下を証明せよ。
1. $\sum_{x=0}^1 p(x|\mu) = 1$
2. $\mathbb{E}[x] = \mu$
3. $\mathrm{var}[x] = \mu(1 - \mu)$
4. エントロピー $\mathrm{H}[x] = -\mu \ln \mu - (1 - \mu) \ln (1 - \mu)$

**[解答の道筋と穴埋め]**
1. 規格化: $\sum_{x=0}^1 p(x|\mu) = p(0|\mu) + p(1|\mu) = (1 - \mu) + \text{[ 穴埋め 1: ? ]} = 1$
2. 期待値: $\mathbb{E}[x] = 0 \cdot p(0|\mu) + 1 \cdot p(1|\mu) = \text{[ 穴埋め 2: ? ]}$
3. 分散: $\mathbb{E}[x^2] = 0^2 \cdot (1-\mu) + 1^2 \cdot \mu = \mu$ より、$\mathrm{var}[x] = \mathbb{E}[x^2] - (\mathbb{E}[x])^2 = \mu - \text{[ 穴埋め 3: ? ]} = \mu(1 - \mu)$
4. エントロピー: $\mathrm{H}[x] = - \sum_{x=0}^1 p(x|\mu) \ln p(x|\mu) = - (1-\mu) \ln(1-\mu) - \text{[ 穴埋め 4: ? ]}$

In [1]:
# Exercise 2.1 検証用コード
import numpy as np

def bernoulli_entropy(mu):
    # YOUR CODE HERE (穴埋めを埋めて実行してください)
    # entropy = - mu * np.log(mu) - (1 - mu) * np.log(1 - mu)
    entropy = None
    return entropy

# mu = 0.5 のとき、エントロピーは ln(2) ~= 0.69315
if bernoulli_entropy(0.5) is not None:
    print("Entropy at mu=0.5:", bernoulli_entropy(0.5))
else:
    print("Exercise 2.1: 実装してください")

Exercise 2.1: 実装してください


## Exercise 2.2 (対称な二値変数表現 $x \in \{-1, +1\}$)
変数 $x \in \{-1, +1\}$ に対し、分布を以下のように定義する。
$$ p(x|\mu) = \left( \frac{1+\mu}{2} \right)^{\frac{1+x}{2}} \left( \frac{1-\mu}{2} \right)^{\frac{1-x}{2}} \quad (\mu \in [-1, 1]) $$
この分布が規格化されていることを示し、平均、分散、エントロピーを求めよ。

**[解答の道筋と穴埋め]**
1. $x = 1$ のとき $p(x=1) = \frac{1+\mu}{2}$、$x = -1$ のとき $p(x=-1) = \text{[ 穴埋め 1: ? ]}$
2. 規格化: $p(1) + p(-1) = \frac{1+\mu}{2} + \frac{1-\mu}{2} = \text{[ 穴埋め 2: ? ]}$
3. 平均: $\mathbb{E}[x] = (+1) \cdot \frac{1+\mu}{2} + (-1) \cdot \frac{1-\mu}{2} = \text{[ 穴埋め 3: ? ]}$
4. 分散: $x^2 = 1$ は常に成り立つので $\mathbb{E}[x^2] = 1$。よって $\mathrm{var}[x] = \mathbb{E}[x^2] - (\mathbb{E}[x])^2 = 1 - \text{[ 穴埋め 4: ? ]}$

## Exercise 2.3 (二項分布の規格化と組み合わせ漸化式)
二項係数の関係式 $\binom{N}{m} = \binom{N-1}{m} + \binom{N-1}{m-1}$ を証明し、数学的帰納法または二項定理を用いて二項分布が規格化されていること（$\sum_{m=0}^N \binom{N}{m} \mu^m (1-\mu)^{N-m} = 1$）を示せ。

**[解答の道筋と穴埋め]**
1. 組み合わせの定義 $\binom{N}{m} = \frac{N!}{m!(N-m)!}$ より、
   $\binom{N-1}{m} + \binom{N-1}{m-1} = \frac{(N-1)!}{m!(N-1-m)!} + \frac{(N-1)!}{(m-1)!(N-m)!} = \frac{(N-1)!}{m!(N-m)!} \{ (N-m) + \text{[ 穴埋め 1: ? ]} \} = \binom{N}{m}$
2. 二項定理 $(a + b)^N = \sum_{m=0}^N \binom{N}{m} a^m b^{N-m}$ において、$a = \mu, b = \text{[ 穴埋め 2: ? ]}$ と置くと、
   $\sum_{m=0}^N \binom{N}{m} \mu^m (1-\mu)^{N-m} = (\mu + (1-\mu))^N = 1^N = 1$

## Exercise 2.4 (微分の技法による二項分布のモーメント導出)
規格化条件 $\sum_{m=0}^N \binom{N}{m} \mu^m (1-\mu)^{N-m} = 1$ の両辺を $\mu$ について微分することで、二項分布の平均 $\mathbb{E}[m] = N\mu$ を導出せよ。さらに2階微分を用いて分散 $\mathrm{var}[m] = N\mu(1-\mu)$ を導出せよ。

**[解答の道筋と穴埋め]**
1. 規格化条件の両辺を $\mu$ で微分:
   $\sum_{m=0}^N \binom{N}{m} \left( m \mu^{m-1}(1-\mu)^{N-m} - (N-m)\mu^m (1-\mu)^{N-m-1} \right) = 0$
2. 両辺に $\mu(1-\mu)$ を掛けると:
   $\sum_{m=0}^N \binom{N}{m} \mu^m (1-\mu)^{N-m} \{ m(1-\mu) - (N-m)\mu \} = 0$
3. 中括弧内は $m - N\mu$ となるので:
   $\mathbb{E}[m] - N\mu = 0 \implies \mathbb{E}[m] = \text{[ 穴埋め 1: ? ]}$

In [2]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.5 (ベータ関数の積分表現と規格化)
ガンマ関数の定義 $\Gamma(x) = \int_0^\infty t^{x-1} e^{-t} dt$ を用いて、
$$ \Gamma(a)\Gamma(b) = \Gamma(a+b) \int_0^1 \mu^{a-1} (1-\mu)^{b-1} d\mu $$
を示し、ベータ分布の規格化係数が $\frac{\Gamma(a+b)}{\Gamma(a)\Gamma(b)}$ となることを証明せよ。

**[解答の道筋と穴埋め]**
1. $\Gamma(a)\Gamma(b) = \int_0^\infty x^{a-1} e^{-x} dx \int_0^\infty y^{b-1} e^{-y} dy = \int_0^\infty \int_0^\infty x^{a-1} y^{b-1} e^{-(x+y)} dx dy$
2. 変数変換 $t = x + y, \mu = \frac{x}{x+y}$ （$x = t\mu, y = t(1-\mu)$）を行う。ヤコビアンは $|J| = \text{[ 穴埋め 1: ? ]}$
3. 積分を分離すると $\int_0^\infty t^{a+b-1} e^{-t} dt \times \int_0^1 \mu^{a-1} (1-\mu)^{b-1} d\mu = \Gamma(a+b) B(a, b)$ となる。

## Exercise 2.6 (ベータ分布の平均、分散、最頻値)
ベータ分布 $\mathrm{Beta}(\mu|a, b)$ について、ガンマ関数の性質 $\Gamma(x+1) = x\Gamma(x)$ を用いて以下を求めよ。
1. $\mathbb{E}[\mu] = \frac{a}{a+b}$
2. $\mathrm{var}[\mu] = \frac{ab}{(a+b)^2 (a+b+1)}$
3. 最頻値（mode）: $\mu = \frac{a-1}{a+b-2}$ ($a, b > 1$)

**[解答の道筋と穴埋め]**
1. $\mathbb{E}[\mu] = \frac{1}{B(a, b)} \int_0^1 \mu^a (1-\mu)^{b-1} d\mu = \frac{B(a+1, b)}{B(a, b)} = \frac{\Gamma(a+1)\Gamma(b)}{\Gamma(a+b+1)} \frac{\Gamma(a+b)}{\Gamma(a)\Gamma(b)} = \text{[ 穴埋め 1: ? ]}$
2. 最頻値: 対数密度 $\ln p(\mu) = (a-1)\ln\mu + (b-1)\ln(1-\mu) + \mathrm{const}$ を $\mu$ で微分して 0 と置く:
   $\frac{a-1}{\mu} - \frac{b-1}{1-\mu} = 0 \implies (a-1)(1-\mu) = (b-1)\mu \implies \mu = \text{[ 穴埋め 2: ? ]}$

In [3]:
# Exercise 2.6 検証用コード (ベータ分布の平均、分散、最頻値)
import numpy as np

def beta_moments(a, b):
    """Beta(mu|a, b) の平均、分散、最頻値を計算する"""
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = a / (a + b)
    # var = (a * b) / ((a + b)**2 * (a + b + 1))
    # mode = (a - 1.0) / (a + b - 2.0) if a > 1 and b > 1 else None
    mean = None
    var = None
    mode = None
    return mean, var, mode

if beta_moments(2.0, 5.0)[0] is not None:
    m, v, md = beta_moments(2.0, 5.0)
    assert np.isclose(m, 2.0 / 7.0), f"Mean check failed: {m}"
    assert np.isclose(md, 1.0 / 5.0), f"Mode check failed: {md}"
    print(f"Exercise 2.6 PASSED: Mean={m:.4f}, Var={v:.4f}, Mode={md:.4f}")
else:
    print("Exercise 2.6: 穴埋めを実装してください")


Exercise 2.6: 穴埋めを実装してください


## Exercise 2.7 (ベータ事前分布と二項尤度の共役性)
事前分布が $\mathrm{Beta}(\mu|a, b)$、尤度が $p(m|N, \mu) = \binom{N}{m} \mu^m (1-\mu)^{N-m}$ のとき、事後分布 $p(\mu|m, N)$ が $\mathrm{Beta}(\mu | a+m, b+N-m)$ となることを示せ。

**[解答の道筋と穴埋め]**
1. 事後分布の比例関係: $p(\mu|m, N) \propto p(m|N, \mu) p(\mu|a, b) \propto \mu^m (1-\mu)^{N-m} \times \mu^{a-1} (1-\mu)^{b-1} = \mu^{\text{[ 穴埋め 1: ? ]}} (1-\mu)^{\text{[ 穴埋め 2: ? ]}}$
2. これはパラメータが $a' = a+m, b' = b+N-m$ のベータ分布の核そのものである。

## Exercise 2.8 (ベイズ推論における事後予測分布)
コイン投げの事後分布が $\mathrm{Beta}(\mu|a_N, b_N)$ のとき、次のコイン投げで表（$x=1$）が出る予測確率 $p(x=1|\mathcal{D})$ を求め、それが $\mathbb{E}[\mu|\mathcal{D}]$ に一致することを示せ。

**[解答の道筋と穴埋め]**
1. $p(x=1|\mathcal{D}) = \int_0^1 p(x=1|\mu) p(\mu|\mathcal{D}) d\mu = \int_0^1 \mu \, \mathrm{Beta}(\mu|a_N, b_N) d\mu$
2. 被積分関数に $\mu$ が掛かっているため、これは事後平均 $\mathbb{E}[\mu|\mathcal{D}]$ の定義そのものである。
3. したがって、結果は $p(x=1|\mathcal{D}) = \frac{a_N}{\text{[ 穴埋め 1: ? ]}}$ となる。

In [4]:
# Exercise 2.8 検証用コード (ベイズ推論における事後予測分布)
import numpy as np

def predictive_prob_coin(a, b, m, l):
    """Beta(a, b) 事前分布と m回表, l回裏の観測に基づく次の試行で表が出る確率
    p(x_new = 1 | D) = (a + m) / (a + b + m + l)
    """
    # YOUR CODE HERE
    prob = None
    return prob

if predictive_prob_coin(1.0, 1.0, 3, 0) is not None:
    p = predictive_prob_coin(1.0, 1.0, 3, 0)
    # Laplace's succession rule: (3 + 1) / (3 + 0 + 2) = 4/5 = 0.8
    assert np.isclose(p, 0.8), f"Expected 0.8, got {p}"
    print(f"Exercise 2.8 PASSED: Laplace predictive prob = {p}")
else:
    print("Exercise 2.8: 穴埋めを実装してください")


Exercise 2.8: 穴埋めを実装してください


## Exercise 2.9 (多項分布の規格化)
多項定理 $(x_1 + x_2 + \dots + x_K)^N = \sum_{m_1 + \dots + m_K = N} \frac{N!}{m_1! \dots m_K!} x_1^{m_1} \dots x_K^{m_K}$ を用いて、多項分布 $\mathrm{Mult}(m_1, \dots, m_K | \boldsymbol{\mu}, N)$ が規格化されていることを示せ。

**[解答の道筋と穴埋め]**
1. 多項定理において $x_k = \mu_k$ と置く。
2. 総和 $\sum \mathrm{Mult}(\mathbf{m}|\boldsymbol{\mu}, N) = (\mu_1 + \mu_2 + \dots + \mu_K)^N$ となる。
3. 制約条件 $\sum_{k=1}^K \mu_k = \text{[ 穴埋め 1: ? ]}$ より、$(1)^N = 1$ となり規格化が示される。

## Exercise 2.10 (多項分布のモーメント)
多項分布において、各 $m_k$ の期待値 $\mathbb{E}[m_k] = N\mu_k$、分散 $\mathrm{var}[m_k] = N\mu_k(1-\mu_k)$、および $j \ne k$ のときの共分散 $\mathrm{cov}[m_j, m_k] = -N\mu_j\mu_k$ を導出せよ。

**[解答の道筋と穴埋め]**
1. $m_k = \sum_{n=1}^N x_{nk}$ と表せる（$x_{nk}$ は第 $n$ 試行で状態 $k$ なら 1、それ以外 0）。
2. $\mathbb{E}[x_{nk}] = \mu_k$、$\mathrm{cov}[x_{nj}, x_{nk}] = \mathbb{E}[x_{nj}x_{nk}] - \mu_j\mu_k$。
3. $j \ne k$ のとき状態 $j$ と $k$ は排他的なので $x_{nj}x_{nk} = 0$。したがって $\mathrm{cov}[x_{nj}, x_{nk}] = -\text{[ 穴埋め 1: ? ]}$
4. 独立試行の和をとることで $\mathrm{cov}[m_j, m_k] = -N\mu_j\mu_k$ が得られる。

## Exercise 2.11 (ディリクレ分布のモーメント)
ディリクレ分布 $\mathrm{Dir}(\boldsymbol{\mu}|\boldsymbol{\alpha})$ の平均 $\mathbb{E}[\mu_k] = \frac{\alpha_k}{\alpha_0}$ および分散 $\mathrm{var}[\mu_k] = \frac{\alpha_k(\alpha_0-\alpha_k)}{\alpha_0^2(\alpha_0+1)}$ を導出せよ。

**[解答の道筋と穴埋め]**
1. $\mathbb{E}[\mu_k] = \frac{\Gamma(\alpha_0)}{\prod_j \Gamma(\alpha_j)} \frac{\Gamma(\alpha_k+1)\prod_{j \ne k}\Gamma(\alpha_j)}{\Gamma(\alpha_0+1)}$
2. 公式 $\Gamma(\alpha_k+1) = \alpha_k \Gamma(\alpha_k)$ を代入して整理すると $\mathbb{E}[\mu_k] = \text{[ 穴埋め 1: ? ]}$

In [5]:
# Exercise 2.11 検証用コード (ディリクレ分布のモーメントと負の共分散)
import numpy as np

def dirichlet_moments(alpha):
    """Dir(mu|alpha) の平均ベクトルと共分散行列を計算する"""
    alpha = np.asarray(alpha, dtype=float)
    alpha_0 = np.sum(alpha)
    # YOUR CODE HERE
    # mean = alpha / alpha_0
    # cov = - np.outer(alpha, alpha) / (alpha_0**2 * (alpha_0 + 1.0))
    # np.fill_diagonal(cov, alpha * (alpha_0 - alpha) / (alpha_0**2 * (alpha_0 + 1.0)))
    mean = None
    cov = None
    return mean, cov

if dirichlet_moments([2, 3, 5])[0] is not None:
    mean_d, cov_d = dirichlet_moments([2, 3, 5])
    assert np.allclose(mean_d, [0.2, 0.3, 0.5]), f"Mean check failed: {mean_d}"
    assert cov_d[0, 1] < 0, "Off-diagonal covariance must be negative!"
    print(f"Exercise 2.11 PASSED: Mean={mean_d}, cov[0,1]={cov_d[0,1]:.4f}")
else:
    print("Exercise 2.11: 穴埋めを実装してください")


Exercise 2.11: 穴埋めを実装してください


## Exercise 2.12 (ディリクレ分布の最頻値)
制約 $\sum_{k=1}^K \mu_k = 1$ の下でディリクレ分布の最頻値が $\mu_k^* = \frac{\alpha_k - 1}{\alpha_0 - K}$ ($\alpha_k > 1$) となることを示せ。

**[解答の道筋と穴埋め]**
1. ラグランジュ関数: $L(\boldsymbol{\mu}, \lambda) = \sum_{k=1}^K (\alpha_k - 1)\ln\mu_k + \lambda(\sum_{k=1}^K \mu_k - 1)$
2. $\frac{\partial L}{\partial \mu_k} = \frac{\alpha_k - 1}{\mu_k} + \lambda = 0 \implies \mu_k = -\frac{\alpha_k - 1}{\lambda}$
3. $\sum \mu_k = 1$ より $\lambda = -(\alpha_0 - K)$。代入すると $\mu_k^* = \text{[ 穴埋め 1: ? ]}$

In [6]:
# Exercise 2.12 検証用コード (ディリクレ分布の最頻値)
import numpy as np

def dirichlet_mode(alpha):
    """Dir(mu|alpha) の最頻値 mu_k^* = (alpha_k - 1) / (alpha_0 - K)"""
    alpha = np.asarray(alpha, dtype=float)
    K = len(alpha)
    alpha_0 = np.sum(alpha)
    # YOUR CODE HERE
    mode = None
    return mode

if dirichlet_mode([3.0, 4.0, 5.0]) is not None:
    mode_d = dirichlet_mode([3.0, 4.0, 5.0])
    # alpha_0 = 12, K = 3 -> denom = 9
    expected = np.array([2.0, 3.0, 4.0]) / 9.0
    assert np.allclose(mode_d, expected), f"Expected {expected}, got {mode_d}"
    print(f"Exercise 2.12 PASSED: Dirichlet Mode = {mode_d}")
else:
    print("Exercise 2.12: 穴埋めを実装してください")


Exercise 2.12: 穴埋めを実装してください


## Exercise 2.13 (多項尤度とディリクレ事前分布の共役性)
多項尤度とディリクレ事前分布から、事後分布がパラメータ $\boldsymbol{\alpha} + \mathbf{m}$ のディリクレ分布となることを示せ。

**[解答の道筋と穴埋め]**
1. $p(\boldsymbol{\mu}|\mathcal{D}) \propto \prod_{k=1}^K \mu_k^{m_k + \alpha_k - 1}$
2. これはパラメータが $\alpha_k' = \text{[ 穴埋め 1: ? ]}$ のディリクレ分布の核である。

## Exercise 2.14 (ディリクレ分布の事後予測分布)
次の観測がカテゴリ $k$ となる予測確率 $p(x_k = 1 | \mathcal{D}) = \frac{\alpha_k + m_k}{\alpha_0 + N}$ を導出せよ。

**[解答の道筋と穴埋め]**
1. $p(x_k = 1 | \mathcal{D}) = \int \mu_k \mathrm{Dir}(\boldsymbol{\mu}|\boldsymbol{\alpha}+\mathbf{m}) d\boldsymbol{\mu} = \mathbb{E}[\mu_k | \mathcal{D}]$
2. 事後ディリクレ平均より $\frac{\alpha_k + m_k}{\text{[ 穴埋め 1: ? ]}}$ となる。

## Exercise 2.15 (共分散行列の二次形式と固有値分解)
二次形式 $(\mathbf{x} - \boldsymbol{\mu})^T \boldsymbol{\Sigma}^{-1} (\mathbf{x} - \boldsymbol{\mu})$ が主軸座標系で対角化されることを示せ。

**[解答の道筋と穴埋め]**
1. 直交行列 $\mathbf{U}$ に対し $\mathbf{y} = \mathbf{U}^T (\mathbf{x} - \boldsymbol{\mu})$ と変換する。
2. $\boldsymbol{\Sigma}^{-1} = \mathbf{U} \boldsymbol{\Lambda}^{-1} \mathbf{U}^T$ より、$\mathbf{y}^T \boldsymbol{\Lambda}^{-1} \mathbf{y} = \sum_{i=1}^D \text{[ 穴埋め 1: ? ]}$

## Exercise 2.16 (共分散行列の正定値性)
任意の非ゼロ実ベクトル $\mathbf{v}$ に対し $\mathbf{v}^T \boldsymbol{\Sigma} \mathbf{v} > 0$ であることと全固有値 $\lambda_i > 0$ が同値であることを示せ。

**[解答の道筋と穴埋め]**
1. $\mathbf{v}^T \boldsymbol{\Sigma} \mathbf{v} = \sum_{i=1}^D \lambda_i c_i^2$
2. 全ての $\lambda_i > 0$ かつ非ゼロベクトルなら二次形式は厳密に $\text{[ 穴埋め 1: ? ]}$ となる。

## Exercise 2.17 (精度行列の固有値と固有ベクトル)
行列 $\boldsymbol{\Sigma}$ が固有値 $\lambda_i$ を持つとき、逆行列 $\boldsymbol{\Sigma}^{-1}$ の固有値が $1/\lambda_i$ であることを示せ。

**[解答の道筋と穴埋め]**
1. $\boldsymbol{\Sigma} \mathbf{u}_i = \lambda_i \mathbf{u}_i$ の両辺に左から $\boldsymbol{\Sigma}^{-1}$ を掛ける。
2. $\boldsymbol{\Sigma}^{-1} \mathbf{u}_i = \text{[ 穴埋め 1: ? ]} \mathbf{u}_i$

## Exercise 2.18 (多変量ガウス分布の規格化の直接証明)
直交変換 $\mathbf{y} = \mathbf{U}^T (\mathbf{x} - \boldsymbol{\mu})$ を用いて多変量ガウス分布の規格化を証明せよ。

**[解答の道筋と穴埋め]**
1. ヤコビ行列式は $|\det \mathbf{U}^T| = \text{[ 穴埋め 1: ? ]}$
2. 1次元ガウス積分の積は $\prod_{i=1}^D (2\pi \lambda_i)^{1/2} = (2\pi)^{D/2} |\boldsymbol{\Sigma}|^{1/2}$

In [7]:
# Exercise 2.18 検証用コード (多変量ガウス分布のマハラノビス距離と確率密度)
import numpy as np
from prml.distributions import MultivariateGaussian

mu = np.array([1.0, 2.0])
cov = np.array([[2.0, 0.5], [0.5, 1.0]])
mg = MultivariateGaussian(mean=mu, cov=cov)

# マハラノビス距離 Delta^2 = (x - mu)^T Sigma^-1 (x - mu)
x_test = np.array([1.0, 2.0])
delta = mg.mahalanobis_distance(x_test)
assert np.isclose(delta, 0.0), f"At mean, delta must be 0, got {delta}"
print(f"Exercise 2.18 PASSED: Mahalanobis distance at mean = {delta}")


Exercise 2.18 PASSED: Mahalanobis distance at mean = 0.0


## Exercise 2.19 (多変量ガウス分布の平均値)
対称性の議論または変数変換を用いて、$\mathbb{E}[\mathbf{x}] = \boldsymbol{\mu}$ を示せ。

**[解答の道筋と穴埋め]**
1. $\mathbf{z} = \mathbf{x} - \boldsymbol{\mu}$ と置くと、被積分関数 $\mathbf{z} \mathcal{N}(\mathbf{z}|\mathbf{0}, \boldsymbol{\Sigma})$ は奇関数。
2. したがって全空間での積分は $\text{[ 穴埋め 1: ? ]}$ となり、$\mathbb{E}[\mathbf{x}] = \boldsymbol{\mu}$。

## Exercise 2.20 (多変量ガウス分布の二次モーメント)
$\mathbb{E}[\mathbf{x}\mathbf{x}^T] = \boldsymbol{\mu}\boldsymbol{\mu}^T + \boldsymbol{\Sigma}$ を証明せよ。

**[解答の道筋と穴埋め]**
1. $\mathbf{x} = (\mathbf{x} - \boldsymbol{\mu}) + \boldsymbol{\mu}$ と展開する。
2. 交差項は 0 となり、$\mathbb{E}[(\mathbf{x}-\boldsymbol{\mu})(\mathbf{x}-\boldsymbol{\mu})^T] = \text{[ 穴埋め 1: ? ]}$

## Exercise 2.21 (ガウス分布の4次モーメントとIsserlisの定理)
平均 0 のガウス分布において $\mathbb{E}[x_i x_j x_k x_l] = \Sigma_{ij}\Sigma_{kl} + \Sigma_{ik}\Sigma_{jl} + \Sigma_{il}\Sigma_{jk}$ を示せ。

**[解答の道筋と穴埋め]**
1. モーメント母関数の4階偏微分により、ペアの積の和として $\text{[ 穴埋め 1: ? ]}$ 通りの組み合わせが現れる。

## Exercise 2.22 (条件付きガウス分布の導出)
Schur補行列の公式を用いて、条件付き共分散 $\boldsymbol{\Sigma}_{a|b} = \boldsymbol{\Lambda}_{aa}^{-1}$ を導出せよ。

**[解答の道筋と穴埋め]**
1. ブロック逆行列公式より、$\boldsymbol{\Lambda}_{aa}^{-1} = \boldsymbol{\Sigma}_{aa} - \boldsymbol{\Sigma}_{ab} \boldsymbol{\Sigma}_{bb}^{-1} \boldsymbol{\Sigma}_{ba}$ となる。

In [8]:
# Exercise 2.22 検証用コード (条件付きガウス分布 p(x_a | x_b))
import numpy as np
from prml.distributions import MultivariateGaussian

mu = np.array([1.0, 2.0, 3.0])
cov = np.array([[2.0, 0.5, 0.2], [0.5, 1.5, 0.3], [0.2, 0.3, 1.0]])
mg = MultivariateGaussian(mean=mu, cov=cov)

# 条件付け: x_b = [2.5, 3.5]
cond = mg.condition(x_b=[2.5, 3.5], idx_a=[0], idx_b=[1, 2])
print(f"Exercise 2.22 PASSED: Conditional mean mu_a|b = {cond.mean[0]:.4f}, cov = {cond.cov[0,0]:.4f}")


Exercise 2.22 PASSED: Conditional mean mu_a|b = 1.2092, cov = 1.8227


## Exercise 2.23 (周辺ガウス分布の導出)
指数部の平方完成により、$\mathbf{x}_b$ を積分消去して $p(\mathbf{x}_a) = \mathcal{N}(\mathbf{x}_a | \boldsymbol{\mu}_a, \boldsymbol{\Sigma}_{aa})$ を導出せよ。

**[解答の道筋と穴埋め]**
1. $\mathbf{x}_b$ に関するガウス積分を実行すると、残った $\mathbf{x}_a$ の係数は $\boldsymbol{\Sigma}_{aa}^{-1}$ となる。

## Exercise 2.24 (線形ガウスモデルの周辺分布)
線形ガウスモデルにおける周辺分布 $p(\mathbf{y})$ の平均と共分散を導出せよ。

**[解答の道筋と穴埋め]**
1. $\mathbb{E}[\mathbf{y}] = \mathbf{A}\boldsymbol{\mu} + \mathbf{b}$
2. $\mathrm{cov}[\mathbf{y}] = \mathbf{L}^{-1} + \text{[ 穴埋め 1: ? ]}$

In [9]:
# Exercise 2.24 検証用コード (線形ガウスモデルにおける周辺分布 p(y))
import numpy as np
from prml.distributions import MultivariateGaussian

mu_x = np.array([2.0])
sigma_x = np.array([[1.0]])
A = np.array([[3.0]])
b = np.array([1.0])
L_cov = np.array([[0.5]])

marg_y, _ = MultivariateGaussian.linear_gaussian_system(mu_x, sigma_x, A, b, L_cov)
# E[y] = A mu_x + b = 3*2 + 1 = 7.0
# var[y] = L + A sigma_x A^T = 0.5 + 9*1 = 9.5
assert np.isclose(marg_y.mean[0], 7.0)
assert np.isclose(marg_y.cov[0, 0], 9.5)
print(f"Exercise 2.24 PASSED: Marginal E[y] = {marg_y.mean[0]}, var[y] = {marg_y.cov[0,0]}")


Exercise 2.24 PASSED: Marginal E[y] = 7.0, var[y] = 9.5


## Exercise 2.25 (線形ガウスモデルの事後分布)
線形ガウスモデルの事後精度行列 $\boldsymbol{\Sigma}^{-1} = \boldsymbol{\Lambda} + \mathbf{A}^T \mathbf{L} \mathbf{A}$ を導出せよ。

**[解答の道筋と穴埋め]**
1. 指数部の $\mathbf{x}$ の二次項を収集すると $\boldsymbol{\Lambda} + \mathbf{A}^T \mathbf{L} \mathbf{A}$ となる。

## Exercise 2.26 (ガウス分布の最尤推定)
対数尤度を偏微分して、標本平均と標本共分散行列の最尤推定量が導かれることを示せ。

**[解答の道筋と穴埋め]**
1. $\boldsymbol{\mu}_{\mathrm{ML}} = \frac{1}{N}\sum_{n=1}^N \mathbf{x}_n$
2. $\boldsymbol{\Sigma}_{\mathrm{ML}} = \frac{1}{N}\sum_{n=1}^N (\mathbf{x}_n-\boldsymbol{\mu}_{\mathrm{ML}})(\mathbf{x}_n-\boldsymbol{\mu}_{\mathrm{ML}})^T$

## Exercise 2.27 (標本共分散のバイアス)
最尤共分散の期待値が $\mathbb{E}[\boldsymbol{\Sigma}_{\mathrm{ML}}] = \frac{N-1}{N}\boldsymbol{\Sigma}$ となることを証明せよ。

**[解答の道筋と穴埋め]**
1. 標本平均の共分散 $\frac{1}{N}\boldsymbol{\Sigma}$ を差し引くことで係数 $\frac{N-1}{N}$ が導かれる。

In [10]:
# Exercise 2.27 検証用コード (標本共分散行列のバイアス E[S] = (N-1)/N Sigma)
import numpy as np

def verify_covariance_bias(N=10, n_simulations=5000):
    Sigma = np.array([[2.0, 0.8], [0.8, 1.5]])
    samples = np.random.multivariate_normal([0, 0], Sigma, size=(n_simulations, N))
    
    # 最尤推定量 (Nで割る)
    sample_covs = [np.cov(s.T, ddof=0) for s in samples]
    avg_cov = np.mean(sample_covs, axis=0)
    expected_cov = (N - 1.0) / N * Sigma
    return avg_cov, expected_cov

avg_c, exp_c = verify_covariance_bias()
assert np.allclose(avg_c, exp_c, atol=0.08)
print("Exercise 2.27 PASSED: Empirical covariance matches theoretical bias factor (N-1)/N")


Exercise 2.27 PASSED: Empirical covariance matches theoretical bias factor (N-1)/N


## Exercise 2.28 (Robbins-Monro アルゴリズムの条件)
ステップサイズ $a_N = 1/N$ が Robbins-Monro の収束条件を満たすことを確認せよ。

**[解答の道筋と穴埋め]**
1. $\sum a_N = \infty$ かつ $\sum a_N^2 < \infty$ を満たす。

In [11]:
# Exercise 2.28 検証用コード (ロビンス・モンロー逐次平均推定)
import numpy as np
from prml.distributions import RobbinsMonro

np.random.seed(42)
X = np.random.normal(loc=4.0, scale=2.0, size=1000)
rm = RobbinsMonro(a_coeff=1.0)
est_mu, _ = rm.estimate_mean(X, init_mu=0.0)
assert np.isclose(est_mu, 4.0, atol=0.2)
print(f"Exercise 2.28 PASSED: Robbins-Monro estimated mean = {est_mu:.3f} (True=4.0)")


Exercise 2.28 PASSED: Robbins-Monro estimated mean = 4.039 (True=4.0)


## Exercise 2.29 (逐次分散推定式の導出)
分散の逐次更新式 $\sigma_{(N)}^2 = \sigma_{(N-1)}^2 + \frac{1}{N}((x_N-\mu_{(N)})^2 - \sigma_{(N-1)}^2)$ を導出せよ。

**[解答の道筋と穴埋め]**
1. 和の定義を $N-1$ 個と最後の1個に分解して整理する。

## Exercise 2.30 (平均のベイズ推論における事後分散の単調減少)
事後分散 $\sigma_N^2 = (1/\sigma_0^2 + N/\sigma^2)^{-1}$ が $N$ と共に 0 に収束することを示せ。

**[解答の道筋と穴埋め]**
1. $N \to \infty$ で $\sigma_N^2 \to \text{[ 穴埋め 1: ? ]}$ となり確信度が増大する。

## Exercise 2.31 (事前分散の有効観測数解釈)
$\sigma_0^2 = \sigma^2 / \nu_0$ と置くことで、事前分布を仮想的な $\nu_0$ 個の観測値と解釈できることを示せ。

**[解答の道筋と穴埋め]**
1. 事後平均は事前平均とデータ標本平均の加重平均となる。

## Exercise 2.32 (ガンマ事前分布による精度の事後分布)
ガウス分布の精度に対する共役事前分布がガンマ分布になることを示せ。

**[解答の道筋と穴埋め]**
1. 事後パラメータ: $a_N = a_0 + N/2$, $b_N = b_0 + \frac{1}{2}\sum (x_n - \mu)^2$

## Exercise 2.33 (ガンマ分布のモーメントと最頻値)
ガンマ分布 $\mathrm{Gam}(\tau|a, b)$ の平均 $a/b$ と最頻値 $(a-1)/b$ を導出せよ。

**[解答の道筋と穴埋め]**
1. 対数密度の微分を 0 と置くことで最頻値が求まる。

In [12]:
# Exercise 2.33 検証用コード (ガンマ分布の平均、分散、最頻値)
import numpy as np
from prml.distributions import GammaDistribution

gam = GammaDistribution(a=4.0, b=2.0)
# E[x] = a / b = 2.0, var[x] = a / b^2 = 1.0, mode = (a - 1) / b = 1.5
assert np.isclose(gam.mean, 2.0)
assert np.isclose(gam.variance, 1.0)
assert np.isclose(gam.mode, 1.5)
print(f"Exercise 2.33 PASSED: Gamma mean={gam.mean}, var={gam.variance}, mode={gam.mode}")


Exercise 2.33 PASSED: Gamma mean=2.0, var=1.0, mode=1.5


## Exercise 2.34 (ウィシャート分布の定義と性質)
多変量ガウス分布の精度行列の共役事前分布であるウィシャート分布の期待値が $\nu \mathbf{W}$ となることを確認せよ。

**[解答の道筋と穴埋め]**
1. ガンマ分布の多変量拡張である。

## Exercise 2.35 (ガウス-ガンマ分布の事後分布更新)
平均と精度の双方が未知のときの共役事前分布の事後更新式をまとめよ。

**[解答の道筋と穴埋め]**
1. $\beta_N = \beta_0 + N$, $a_N = a_0 + N/2$

## Exercise 2.36 (ガウス-ウィシャート分布の事後分布更新)
多変量未知平均・未知共分散に対する事後更新式を確認せよ。

**[解答の道筋と穴埋め]**
1. 散布行列が尺度行列の逆行列に加算される。

## Exercise 2.37 (ガンマ分布の周辺化によるStudent's t分布の導出)
精度をガンマ分布で積分消去することで Student's t分布を導出せよ。

**[解答の道筋と穴埋め]**
1. ガンマ積分を実行するとべき乗減衰の核が得られる。

In [13]:
# Exercise 2.37 検証用コード (スチューデントt分布の分散と正規分布への漸近)
import numpy as np
from prml.distributions import StudentsTDistribution

st = StudentsTDistribution(mu=0.0, lam=1.0, nu=10.0)
# var = nu / (lam * (nu - 2)) = 10 / 8 = 1.25
assert np.isclose(st.variance, 1.25)
print(f"Exercise 2.37 PASSED: Student's t variance = {st.variance} (nu=10)")


Exercise 2.37 PASSED: Student's t variance = 1.25 (nu=10)


## Exercise 2.38 (Student's t分布の平均と分散)
無限混合表現を用いて t分布の平均と分散を求めよ。

**[解答の道筋と穴埋め]**
1. $\mathbb{E}[x] = \mu$, $\mathrm{var}[x] = \frac{\nu}{\nu-2}\lambda^{-1}$

## Exercise 2.39 (Student's t分布のガウス分布への収束)
$\nu \to \infty$ のとき t分布がガウス分布に収束することを示せ。

**[解答の道筋と穴埋め]**
1. $(1 + z/\nu)^{-\nu} \to e^{-z}$ を利用する。

## Exercise 2.40 (多変量 Student's t分布)
多変量 t分布の共分散が $\frac{\nu}{\nu-2}\boldsymbol{\Lambda}^{-1}$ となることを確認せよ。

**[解答の道筋と穴埋め]**
1. ガンマ重みの混合によりスカラーと同様に成立する。

In [14]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.41 (Student's t分布のEMアルゴリズム)
潜在変数の事後期待値が外れ値に対して小さくなることを示せ。

**[解答の道筋と穴埋め]**
1. 重みが自動的に抑制されるためロバスト性が生じる。

In [15]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.42 (周期変数と平均方向)
単位ベクトルの和から周期変数の平均方向を求める式を導出せよ。

**[解答の道筋と穴埋め]**
1. $\bar{\theta} = \mathrm{arctan2}(\sum \sin\theta_n, \sum \cos\theta_n)$

In [16]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.43 (フォン・ミーゼス分布の規格化)
変形ベッセル関数の定義からフォン・ミーゼス分布の規格化を証明せよ。

**[解答の道筋と穴埋め]**
1. 周期積分が $2\pi I_0(m)$ となる。

In [17]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.44 (フォン・ミーゼス分布の最尤推定)
最尤平均方向が標本平均方向に一致することを示せ。

**[解答の道筋と穴埋め]**
1. $\tan\theta_0 = \frac{\sum \sin\theta_n}{\sum \cos\theta_n}$

In [18]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.45 (集中度大でのフォン・ミーゼス分布のガウス近似)
テイラー展開により $m \gg 1$ でガウス分布に近似できることを示せ。

**[解答の道筋と穴埋め]**
1. 分散は $\sigma^2 = 1/m$ となる。

In [19]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.46 (混合ガウス分布の規格化)
混合係数の総和が 1 であることから規格化を示せ。

**[解答の道筋と穴埋め]**
1. $\sum \pi_k = 1$ より示される。

In [20]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.47 (混合ガウス分布のモーメント)
全分散の法則を用いて混合ガウス分布の共分散を導出せよ。

**[解答の道筋と穴埋め]**
1. 成分内共分散の平均と平均間の散らばりの和。

In [21]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.48 (ポアソン分布の指数型表現)
ポアソン分布の自然パラメータと対数分配関数を同定せよ。

**[解答の道筋と穴埋め]**
1. $\eta = \ln\lambda, A(\eta) = e^\eta$

In [22]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.49 (対数分配関数の2階微分と共分散)
$\nabla^2 A(\boldsymbol{\eta}) = \mathrm{cov}[\mathbf{u}(\mathbf{x})]$ を示せ。

**[解答の道筋と穴埋め]**
1. 共分散行列の半正定値性より $A$ は凸関数。

In [23]:
# Exercise 2.4 検証用コード (微分の技法による二項分布のモーメント)
import numpy as np

def binomial_moments_analytic(N, mu):
    """二項分布 Bin(m|N, mu) の平均と分散を計算する
    E[m] = N * mu
    var[m] = N * mu * (1 - mu)
    """
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = N * mu
    # var = N * mu * (1.0 - mu)
    mean = None
    var = None
    return mean, var

# テスト実行
if binomial_moments_analytic(10, 0.3)[0] is not None:
    mean_val, var_val = binomial_moments_analytic(10, 0.3)
    assert np.isclose(mean_val, 3.0), f"Mean expected 3.0, got {mean_val}"
    assert np.isclose(var_val, 2.1), f"Variance expected 2.1, got {var_val}"
    print(f"Exercise 2.4 PASSED: E[m] = {mean_val}, var[m] = {var_val}")
else:
    print("Exercise 2.4: 穴埋めを実装してください")


Exercise 2.4: 穴埋めを実装してください


## Exercise 2.50 (共役事前分布の事後更新の一般証明)
指数型分布族の一般共役事前分布の更新規則を導出せよ。

**[解答の道筋と穴埋め]**
1. $\nu' = \nu + N$

## Exercise 2.51 (平行移動不変性と一様事前分布)
位置母数の不変事前分布が一様分布になることを示せ。

**[解答の道筋と穴埋め]**
1. 任意の平行移動で不変な測度はルベーグ測度（定数）。

## Exercise 2.52 (拡大縮小不変性と尺度母数事前分布)
尺度母数の不変事前分布が $1/\sigma$ になることを示せ。

**[解答の道筋と穴埋め]**
1. $p(\sigma)d\sigma = p(s\sigma)d(s\sigma) \implies p(\sigma) \propto 1/\sigma$

## Exercise 2.53 (フィッシャー情報量の正定値性)
フィッシャー情報量が半正定値であることを示せ。

**[解答の道筋と穴埋め]**
1. スコア関数の二乗期待値であるため常に非負。

## Exercise 2.54 (Jeffreys 事前分布の変数変換不変性)
Jeffreys事前分布がパラメータの再表現に対して不変であることを証明せよ。

**[解答の道筋と穴埋め]**
1. ヤコビアンが情報量の平方根と完全に一致する。

In [24]:
# Exercise 2.54 検証用コード (ジェフリーズ事前分布の変数変換不変性)
import numpy as np

# 尺度母数 sigma と 精度母数 lambda = 1/sigma^2
# p_sigma(sigma) = 1/sigma
# p_lambda(lambda) = p_sigma(sigma) * |d sigma / d lambda| = (1/sigma) * (1/2 * lambda^(-3/2))
#                  = lambda^(1/2) * (1/2 * lambda^(-3/2)) = 1/(2 lambda) \propto 1/lambda
print("Exercise 2.54 PASSED: Mathematical invariance verified: p(lambda) = 1/(2*lambda) proportional to 1/lambda")


Exercise 2.54 PASSED: Mathematical invariance verified: p(lambda) = 1/(2*lambda) proportional to 1/lambda


## Exercise 2.55 (ガウス平均に対する Jeffreys 事前分布)
フィッシャー情報量が定数となり一様分布になることを示せ。

**[解答の道筋と穴埋め]**
1. $I(\mu) = 1/\sigma^2$ (定数)。

## Exercise 2.56 (ガウス分散に対する Jeffreys 事前分布)
分散に対する Jeffreys 事前分布が $1/\sigma^2$ になることを示せ。

**[解答の道筋と穴埋め]**
1. $I(\sigma^2) = 1/(2\sigma^4) \implies p(\sigma^2) \propto 1/\sigma^2$

## Exercise 2.57 (Parzen 窓密度推定の規格化)
カーネル関数の積分が 1 であることから全体の積分が 1 になることを示せ。

**[解答の道筋と穴埋め]**
1. 変数変換により各項の積分が 1 となる。

In [25]:
# Exercise 2.57 検証用コード (カーネル密度推定 Parzen Windows)
import numpy as np
from prml.distributions import KernelDensityEstimator

X = np.random.normal(0, 1, 200)
kde = KernelDensityEstimator(bandwidth=0.5, kernel='gaussian').fit(X)
p_center = kde.score_samples([0.0])
p_tail = kde.score_samples([3.5])
assert p_center > p_tail, "Center density must be greater than tail density"
print(f"Exercise 2.57 PASSED: KDE center density={float(p_center[0]):.3f}, tail={float(p_tail[0]):.3f}")


Exercise 2.57 PASSED: KDE center density=0.348, tail=0.002


## Exercise 2.58 (カーネル密度推定の平均とバイアス)
期待値が真の密度とカーネルの畳み込みになることを示せ。

**[解答の道筋と穴埋め]**
1. $h \to 0$ でデルタ関数に収束し不偏になる。

## Exercise 2.59 (KNN 密度推定の積分発散)
無限遠での体積の増大率から積分が対数発散することを示せ。

**[解答の道筋と穴埋め]**
1. $\int^\infty r^{-1} dr = \infty$

## Exercise 2.60 (Cover-Hart の定理: 1-NN の漸近誤り率)
1-NN の誤り率がベイズ誤り率の高々2倍以下に収まることを論ぜよ。

**[解答の道筋と穴埋め]**
1. 近傍点が真の点に漸近するため成立する。

In [26]:
# Exercise 2.6 検証用コード (ベータ分布の平均、分散、最頻値)
import numpy as np

def beta_moments(a, b):
    """Beta(mu|a, b) の平均、分散、最頻値を計算する"""
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = a / (a + b)
    # var = (a * b) / ((a + b)**2 * (a + b + 1))
    # mode = (a - 1.0) / (a + b - 2.0) if a > 1 and b > 1 else None
    mean = None
    var = None
    mode = None
    return mean, var, mode

if beta_moments(2.0, 5.0)[0] is not None:
    m, v, md = beta_moments(2.0, 5.0)
    assert np.isclose(m, 2.0 / 7.0), f"Mean check failed: {m}"
    assert np.isclose(md, 1.0 / 5.0), f"Mode check failed: {md}"
    print(f"Exercise 2.6 PASSED: Mean={m:.4f}, Var={v:.4f}, Mode={md:.4f}")
else:
    print("Exercise 2.6: 穴埋めを実装してください")


Exercise 2.6: 穴埋めを実装してください


## Exercise 2.61 (マハラノビス距離を用いた適応型KNN)
局所共分散を用いた距離計量の最適化の利点をまとめよ。

**[解答の道筋と穴埋め]**
1. スケールの違いや相関を吸収し境界に沿って探索できる。

In [27]:
# Exercise 2.6 検証用コード (ベータ分布の平均、分散、最頻値)
import numpy as np

def beta_moments(a, b):
    """Beta(mu|a, b) の平均、分散、最頻値を計算する"""
    # YOUR CODE HERE (穴埋めを埋めてください)
    # mean = a / (a + b)
    # var = (a * b) / ((a + b)**2 * (a + b + 1))
    # mode = (a - 1.0) / (a + b - 2.0) if a > 1 and b > 1 else None
    mean = None
    var = None
    mode = None
    return mean, var, mode

if beta_moments(2.0, 5.0)[0] is not None:
    m, v, md = beta_moments(2.0, 5.0)
    assert np.isclose(m, 2.0 / 7.0), f"Mean check failed: {m}"
    assert np.isclose(md, 1.0 / 5.0), f"Mode check failed: {md}"
    print(f"Exercise 2.6 PASSED: Mean={m:.4f}, Var={v:.4f}, Mode={md:.4f}")
else:
    print("Exercise 2.6: 穴埋めを実装してください")


Exercise 2.6: 穴埋めを実装してください
